docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+}
+```
+The interface is quite similar to the Go [collate package](https://godoc.org/golang.org/x/text/collate), it is possible to implement some of the MySQL collation by simply mapping it to an existing Go collation.
+
+### Row Format
+
+The encoding layout of TiDB has been described in our [previous article](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map). The row format should be changed to make it memory comparable, this is important to the index lookup. Basic principle is that all keys encoded should use the `sortKeys` result from `Key()`/`KeyFromString()` function. However, most of the `sortKeys` calculations are not reversible.
+
+#### Option 1
+
+  * For table data, encodings stay unchanged. All strings are compared after decoding with the `Compare()` function.
+  * For table indices, we replace current `ColumnValue` with `sortKey`:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID
+	```
+    - For non-unique indices:
+    ```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: null
+	```
+
+Pros:
+    The current encoding layout keeps unchanged.
+
+Cons:
+     Since index value can not be recovered from `sortKey`, extra table lookup is needed for current Index-only queries, this may lead to performance regressions. The extra table lookup also requires corresponding changes to the TiDB optimizer.
+
+#### Option 2
+
+  * Table data encodings is the same with Option 1.
+  * For table indices, we encode the `ColumnValue` to the value of the storage, so as to avoid extra table lookup for index queries:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID_indexedColumnsValue
+	```
+    - For non-unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: indexedColumnsValue
+	```
+
+Pros: Extra table lookup is avoided and no change to the TIDB optimizer.
+
+Cons: The size of index value on string column with new collations is doubled(we need to write both `sortKey` and `indexedColumnsValue`, their sizes are equal for most of the collations).
+
+## Compatibility
+
+### Compatibility with MySQL
+
+As stated before, MySQL has various different collation implementations. For example, in MySQL 5.7, the default collation of `utf8mb4` is `utf8mb4_general_ci`, which has two main flaws:
+  * It only implement the `simple mappings` of UCA.
+  * It used a very old version of Unicode.
+In MySQL 8, the default collation of `utf8mb4` has been changed to `utf8mb4_900_ai_ci`, in which the Unicode version is upgraded to 9.0, with full UCA implementation.
+
+Should TiDB aim to support all collation of MySQL?
+  * If TiDB supports the latest collations only(for example, support `utf8mb4_900_ai_ci` only and treat `utf8mb4_general_ci` as one of its alias). This seems to be 'standard-correct' but may lead to potential compatibility issues.
```

Comment:
I would recommend this path. Keep in mind why MySQL can not do this:

- The collation affects the on-disk format. It can't just change meta data to upgrade users to a new version, it needs to upgrade the tables (collations are effectively a data type on indexed values).
- The upgrade can not be guaranteed to be unattended. Due to both bugs, and differences unique key violations may occur.

These will make migrations of edge-cases harder, but applications rarely depend on the behaviour of the edge-cases.  Bug-for-bug compatibility in datatypes is really hardwork, so I suspect even with best intentions (and all collations implemented) this case will still exist where unique constraints errors will be triggered during migrations incorrectly.

An automated way for the DM framwork to handle this, would be to defer creating unique indexes, and first scan and report any violations. They are usually small enough that an operator can fix them - and may want to in the source database.

Reply:
Thanks for the recommendations, especially the idea of uniqueness-detect for DM framework. 

My only concern is that some behaviors may not be detected by the on-storage data scans. For example, the potential changes of application queries, but that should be fine...as you mentioned.

Reply:
Since many of TiDB's users expect to migrate from MySQL without any upper-level change, the basic principle is that TiDB provides as much compatibility as possible with MySQL, which is, most of the time we have to pay the price for the historical burden of MySQL(like `utf8` and `utf8mb4`).

Reply:
I don't want `general_ci`, and after reading the doc of Oracle I think we could do better than MySQL.
The reason that I don't want `general_ci` is:
https://stackoverflow.com/questions/766809/whats-the-difference-between-utf8-general-ci-and-utf8-unicode-ci
and
https://stackoverflow.com/questions/1036454/what-are-the-differences-between-utf8-general-ci-and-utf8-unicode-ci